### ISSUE TITLE
Game crashes when unit enters a city (Sim ruleset)

### ISSUE THREAD

**Issue by Corbeau-1** [longturn/freeciv21#2351]
To reproduce the bug: 

1. Load the savegame with the ruleset provided, both in the attachment below.
2. /take Fidel
3. Find the Raider unit # 1163, just west of main Cuban territory, sitting south of Venetian city of Famagosta
4. Attack/enter Famagosta (occupychance is set to 100)

UPDATE: it seems to happen in **some** instances where there is no city present, just combat. A wild guess: it may have something to do with limited combat rounds and entry into tile after victory.

Using Windows installation 3.1-beta.1.

(updated attachment below)

**Comment by lmoureaux** [comment]
The ruleset doesn't load:
```
[critical] freeciv21-server (/home/louis/Personal/freeciv21/common/requirements.cpp:1086) - Invalid requirement DiplRel | Local |  | present | Is Foreign: bad type or name
[critical] freeciv21-server (/home/louis/Personal/freeciv21/server/ruleset.cpp:453) - "/tmp/abugabug/SIM30/game.ruleset" [actionenabler_investigate_city] has invalid or unknown req: "DiplRel" "Is Foreign".
```

**Comment by lmoureaux** [comment]
Managed to start the game after upgrading the ruleset with `freeciv21-ruleup`. Couldn't reproduce the crash so far, the city gets wiped out and the raider gets promoted.

**Comment by lmoureaux** [comment]
ASan sees it:
```
=================================================================
==2154558==ERROR: AddressSanitizer: heap-use-after-free on address 0x61e00006e4e8 at pc 0x650a0c4aa898 bp 0x7ffd46e4e000 sp 0x7ffd46e4dff0
READ of size 4 at 0x61e00006e4e8 thread T0
    #0 0x650a0c4aa897 in governor::run() /home/louis/Personal/freeciv21/client/governor.cpp:168
    #1 0x650a0c2a9930 in governor::unfreeze() /home/louis/Personal/freeciv21/client/governor.h:26
    #2 0x650a0c2a9930 in input_from_server(QIODevice*) /home/louis/Personal/freeciv21/client/clinet.cpp:325
    #3 0x650a0c445c0f in QtPrivate::FunctorCall<QtPrivate::IndexesList<>, QtPrivate::List<>, void, void (fc_client::*)()>::call(void (fc_client::*)(), fc_client*, void**) /usr/include/x86_64-linux-gnu/qt5/QtCore/qobjectdefs_impl.h:152
    #4 0x650a0c445c0f in void QtPrivate::FunctionPointer<void (fc_client::*)()>::call<QtPrivate::List<>, void>(void (fc_client::*)(), fc_client*, void**) /usr/include/x86_64-linux-gnu/qt5/QtCore/qobjectdefs_impl.h:185
    #5 0x650a0c445c0f in QtPrivate::QSlotObject<void (fc_client::*)(), QtPrivate::List<>, void>::impl(int, QtPrivate::QSlotObjectBase*, QObject*, void**, bool*) /usr/include/x86_64-linux-gnu/qt5/QtCore/qobjectdefs_impl.h:418
    #6 0x650a0c445c0f in QtPrivate::QSlotObject<void (fc_client::*)(), QtPrivate::List<>, void>::impl(int, QtPrivate::QSlotObjectBase*, QObject*, void**, bool*) /usr/include/x86_64-linux-gnu/qt5/QtCore/qobjectdefs_impl.h:411
    #7 0x7f5d4e8f3ed3 in QtPrivate::QSlotObjectBase::call(QObject*, void**) ../../include/QtCore/../../src/corelib/kernel/qobjectdefs_impl.h:398
    #8 0x7f5d4e8f3ed3 in void doActivate<false>(QObject*, int, void**) kernel/qobject.cpp:3925
    #9 0x7f5d4fd0c84e  (/lib/x86_64-linux-gnu/libQt5Network.so.5+0xee84e)
    #10 0x7f5d4fd1f988  (/lib/x86_64-linux-gnu/libQt5Network.so.5+0x101988)
    #11 0x7f5d4f56c6e2 in QApplicationPrivate::notify_helper(QObject*, QEvent*) kernel/qapplication.cpp:3640
    #12 0x7f5d4e8bc429 in QCoreApplication::notifyInternal2(QObject*, QEvent*) kernel/qcoreapplication.cpp:1064
    #13 0x7f5d4e916524 in socketNotifierSourceDispatch kernel/qeventdispatcher_glib.cpp:107
    #14 0x7f5d4c920d3a in g_main_context_dispatch (/lib/x86_64-linux-gnu/libglib-2.0.so.0+0x55d3a)
    #15 0x7f5d4c9762b7  (/lib/x86_64-linux-gnu/libglib-2.0.so.0+0xab2b7)
    #16 0x7f5d4c91e3e2 in g_main_context_iteration (/lib/x86_64-linux-gnu/libglib-2.0.so.0+0x533e2)
    #17 0x7f5d4e915917 in QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) kernel/qeventdispatcher_glib.cpp:423
    #18 0x7f5d4e8bad4a in QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) kernel/qeventloop.cpp:235
    #19 0x7f5d4e8c32f3 in QCoreApplication::exec() kernel/qcoreapplication.cpp:1375
    #20 0x650a0c40dedb in fc_client::fc_main(QApplication*) /home/louis/Personal/freeciv21/client/fc_client.cpp:158
    #21 0x650a0c4bb65c in ui_main() /home/louis/Personal/freeciv21/client/gui_main.cpp:85
    #22 0x650a0c27fc2e in client_main(int, char**) /home/louis/Personal/freeciv21/client/client_main.cpp:611
    #23 0x7f5d4d629d8f in __libc_start_call_main ../sysdeps/nptl/libc_start_call_main.h:58
    #24 0x7f5d4d629e3f in __libc_start_main_impl ../csu/libc-start.c:392
    #25 0x650a0bff7d94 in _start (/home/louis/Personal/freeciv21/build-asan/freeciv21-client+0x2a6bd94)

0x61e00006e4e8 is located 104 bytes inside of 2456-byte region [0x61e00006e480,0x61e00006ee18)
freed by thread T0 here:
    #0 0x7f5d500b6e57 in operator delete[](void*) ../../../../src/libsanitizer/asan/asan_new_delete.cpp:163
    #1 0x650a0d25fcdc in game_remove_city(world*, city*) /home/louis/Personal/freeciv21/common/game.cpp:195

previously allocated by thread T0 here:
    #0 0x7f5d500b6357 in operator new[](unsigned long) ../../../../src/libsanitizer/asan/asan_new_delete.cpp:102
    #1 0x650a0d206ce8 in create_city_virtual(player*, tile*, char const*) /home/louis/Personal/freeciv21/common/city.cpp:3334

SUMMARY: AddressSanitizer: heap-use-after-free /home/louis/Personal/freeciv21/client/governor.cpp:168 in governor::run()
Shadow bytes around the buggy address:
  0x0c3c80005c40: 00 00 00 fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c3c80005c50: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c3c80005c60: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c3c80005c70: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c3c80005c80: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
=>0x0c3c80005c90: fd fd fd fd fd fd fd fd fd fd fd fd fd[fd]fd fd
  0x0c3c80005ca0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c3c80005cb0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c3c80005cc0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c3c80005cd0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c3c80005ce0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==2154558==ABORTING
```

**Comment by Corbeau-1** [comment]
> [actionenabler_investigate_city] has invalid or unknown req: "DiplRel" "Is Foreign".

Er... What?

![image](https://github.com/user-attachments/assets/c0353da0-7c64-4637-a789-0d7edb1c6a40)

**Comment by Corbeau-1** [comment]
This is **in the zip I `supplied**:
`[actionenabler_3]
action="Investigate City"
actor_reqs={"type","name","range","present"
"UnitFlag","Diplomat","Local",TRUE
"MinMoveFrags","1","Local",TRUE
"DiplRel","Foreign","Local",TRUE
"UnitFlag","Spy","Local",FALSE
`}`

Is it possible you loaded the wrong ruleset (and .serv)? Do you have older versions of Sim on your computer?

**Comment by lmoureaux** [comment]
Oh I see that you updated the description. Hadn't noticed until now so I'm still using the "old" one. Anyway the ASan log above is likely pointing at the origin of the crash.

**Comment by Corbeau-1** [comment]
Anyway, I found some other errors in the ruleset, unrelated to this, fixed them, this is the newer version, along with the save.

[SIM30-fc21.zip](https://github.com/user-attachments/files/16838719/SIM30-fc21.zip)

**Comment by Corbeau-1** [comment]
> Oh I see that you updated the description. Hadn't noticed until now so I'm still using the "old" one. Anyway the ASan log above is likely pointing at the origin of the crash.

Actually, I don't know where your thing came from. I've been using the ruleupped ruleset for ages. That part wasn't updated a few days ago.

Anyway, I don't understand this ASan log part. 

And now, with fixed ruleset, the game crashes after an attack that doesn't kill the defending unit so the attacker doesn't enter the city. Just: combat, attacker lost 2 HP, defender lost 8 HP, defender survived, crash.

**Comment by lmoureaux** [comment]
> Anyway, I don't understand this ASan log part.

You're not expected to, it's developer information. I pasted it here to investigate later.

> And now, with fixed ruleset, the game crashes after an attack that doesn't kill the defending unit so the attacker doesn't enter the city. Just: combat, attacker lost 2 HP, defender lost 8 HP, defender survived, one-second freeze, crash.

This looks like a different bug -> new issue. Do you have also a save to reproduce that one easily?

**Comment by Corbeau-1** [comment]
> This looks like a different bug -> new issue. Do you have also a save to reproduce that one easily?

Same save. The previous one had a bug (in relation to Sim): combat until death. This caused the defender in city to die and the attacking unit to enter the city.

The fixed version has combat rounds fixed, it's the same two units, but now defender in the city survives so there is no entry into city. Game still crashes, though.

Interesting thing: further east, I attacked a Worker with a Raider and game didn't crash.

**Comment by lmoureaux** [comment]
Ok thanks. So I should use the now-removed attachment [abugabug2.zip](https://github.com/user-attachments/files/16787222/abugabug2.zip) to investigate the first bug (fortunately it's still available from message history) and [SIM30-fc21.zip](https://github.com/user-attachments/files/16838719/SIM30-fc21.zip) to investigate the second one. Because well, both bugs should be fixed, not just the one affecting you right now :wink:

**Comment by Corbeau-1** [comment]
Hmmm... Right, sorry. Well, I feel it's the same bug, but I could be wrong. 

For the record, on rare occasions the game didn't crash and continued with the unit destroying the city - and also with no-kill combat, so in both versions. But crashes happened more often.

**Comment by lmoureaux** [comment]
I think they're different because the dev information above points to the governor code. This shouldn't be involved in a combat where none of the units dies. (Except that this is Freeciv so...)

**Comment by Corbeau-1** [comment]
New info: I started a new game and there was a lot of combat, no crash, also I took one empty city, but the first time I attacked a defended city, killed a unit inside and automatically entered it, the game crashed.

But if it's about the goverrner, then `occupychance` probably has nothing to do with it, right?

**Comment by Corbeau-1** [comment]
New info. Removed `occupychance` (set it to 0), still crashes. Also, can confirm that game crashes after attacking a city with defender surviving, attacker not entering the city. It **seems** the initial report was due to defender dying and occupier entering the city with the crash taking one second delay to happen, so it was likely caused by combat, not entering the city.

Combat in the field doesn't cause crashing.

**Comment by Corbeau-1** [comment]
Confirmed what I said on Discord: starting the game from server - no crash. Saved, loaded only using the client - crash.

**Comment by lmoureaux** [comment]
First version of this issue -> #2385.

**Comment by lmoureaux** [comment]
Is this happening in Sim06?

**Comment by lmoureaux** [comment]
The new zip contains a savegame with a single opponent that you haven't yet met. I used the save from the first version and attacked the city as instructed in the first post. I couldn't crash the client (not that I tried *very* hard either).

Since I don't hear complaints about crashes in Sim06 and there has been loads of fighting I'm going to close this as completed. The first version of this issue has most certainly been fixed. Open a new issue (possibly referencing this one) if there is still a crash in normal combat outside cities.
